File: nameconstraints/testsuites/JavaTest.java
/**
 *
 *  Copyright 2017 Netflix, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.KeyStore;
import java.security.cert.Certificate;
import java.security.cert.CertificateFactory;

public class JavaTest {

    public static void main(String[] args) throws Exception {
        if (args.length != 2) {
            throw new IllegalArgumentException("Expects two args: dnsUrl ipUrl");
        }
        String dnsUrl = args[0];
        String ipUrl = args[1];

        CertificateFactory certificateFactory = CertificateFactory.getInstance("X509");
        Certificate rootCrt;
        try (InputStream inputStream = Files.newInputStream(Paths.get("../../docs/root.crt"))) {
            rootCrt = certificateFactory.generateCertificate(inputStream);
        }
        KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
        trustStore.load(null, null);
        trustStore.setCertificateEntry("1", rootCrt);
        TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
        tmf.init(trustStore);

        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, tmf.getTrustManagers(), null);

        boolean dnsResult = testUrl(sslContext, dnsUrl);
        boolean ipResult = testUrl(sslContext, ipUrl);

        System.out.println(String.format("{\"dnsResult\":%s, \"ipResult\":%s}", Boolean.toString(dnsResult), Boolean.toString(ipResult)));
    }

    private static boolean testUrl(SSLContext sslContext, String url) {
        try {
            HttpsURLConnection urlConnection = (HttpsURLConnection) new URL(url).openConnection();
            urlConnection.setSSLSocketFactory(sslContext.getSocketFactory());

            int bytesRead = 0;
            try (InputStream inputStream = urlConnection.getInputStream()) {
                byte[] buffer = new byte[4096];
                bytesRead = inputStream.read(buffer);
            }

            return bytesRead > 0;
        } catch (IOException e) {
            return false;
        }
    }
}

File: pathbuilding/impltests/java_test.go
package impltests

import (
	"fmt"
	"io/ioutil"
	"os/exec"
	"path/filepath"
	"testing"
)

func TestJava(t *testing.T) {
	version, err := execAndCapture("java", "-version")
	if err != nil {
		t.Fatal(err)
	}
	t.Log(version)

	tmpDir := t.TempDir()
	err = ioutil.WriteFile(filepath.Join(tmpDir, "Curl.java"), []byte(`
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SNIHostName;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLParameters;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManagerFactory;
import java.io.IOException;
import java.io.InputStream;
import java.net.InetAddress;
import java.net.Socket;
import java.net.URL;
import java.net.UnknownHostException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.KeyStore;
import java.security.cert.Certificate;
import java.security.cert.CertificateFactory;
import java.util.Collections;

public class Curl {
    public static void main(String[] args) throws Exception {
        String caPath = args[0];
        String servername = args[1];
        String url = args[2];

        Certificate rootCert;
        try (InputStream inputStream = Files.newInputStream(Paths.get(caPath))) {
            rootCert = CertificateFactory.getInstance("X509").generateCertificate(inputStream);
        }

        KeyStore truststore = KeyStore.getInstance(KeyStore.getDefaultType());
        truststore.load(null, null);
        truststore.setCertificateEntry("1", rootCert);
        TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
        tmf.init(truststore);

        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, tmf.getTrustManagers(), null);

        HttpsURLConnection connection = (HttpsURLConnection) new URL(url).openConnection();
        connection.setSSLSocketFactory(new SSLSocketFactoryWrapper(sslContext.getSocketFactory(), servername));
        connection.setHostnameVerifier((hostname, session) -> true);
        if (connection.getResponseCode() != 200) {
            throw new AssertionError("Invalid response code: " + connection.getResponseCode());
        }
    }

    private static class SSLSocketFactoryWrapper extends SSLSocketFactory {

        private final SSLSocketFactory wrappedFactory;
        private final String servername;

        public SSLSocketFactoryWrapper(SSLSocketFactory factory, String servername) {
            this.wrappedFactory = factory;
            this.servername = servername;
        }

        @Override
        public Socket createSocket(String host, int port) throws IOException {
            SSLSocket socket = (SSLSocket) wrappedFactory.createSocket(host, port);
            setParameters(socket);
            return socket;
        }

        @Override
        public Socket createSocket(String host, int port, InetAddress localHost, int localPort) throws IOException {
            SSLSocket socket = (SSLSocket) wrappedFactory.createSocket(host, port, localHost, localPort);
            setParameters(socket);
            return socket;
        }


        @Override
        public Socket createSocket(InetAddress host, int port) throws IOException {
            SSLSocket socket = (SSLSocket) wrappedFactory.createSocket(host, port);
            setParameters(socket);
            return socket;
        }

        @Override
        public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort) throws IOException {
            SSLSocket socket = (SSLSocket) wrappedFactory.createSocket(address, port, localAddress, localPort);
            setParameters(socket);
            return socket;

        }

        @Override
        public Socket createSocket() throws IOException {
            SSLSocket socket = (SSLSocket) wrappedFactory.createSocket();
            setParameters(socket);
            return socket;
        }

        @Override
        public String[] getDefaultCipherSuites() {
            return wrappedFactory.getDefaultCipherSuites();
        }

        @Override
        public String[] getSupportedCipherSuites() {
            return wrappedFactory.getSupportedCipherSuites();
        }

        @Override
        public Socket createSocket(Socket s, String host, int port, boolean autoClose) throws IOException {
            SSLSocket socket = (SSLSocket) wrappedFactory.createSocket(s, host, port, autoClose);
            setParameters(socket);
            return socket;
        }

        private void setParameters(SSLSocket socket) {
            SSLParameters parameters = socket.getSSLParameters();
            parameters.setServerNames(Collections.singletonList(new SNIHostName(servername)));
            socket.setSSLParameters(parameters);
        }
    }
}
`), 0644)
	if err != nil {
		t.Fatal(err)
	}
	cmd := exec.Command("javac", "Curl.java")
	cmd.Dir = tmpDir
	err = cmd.Run()
	if err != nil {
		t.Fatal(err)
	}

	testExec(t, func(caPath string, testCaseName string, tlsPort int) []string {
		return []string{
			"java", "-Djdk.tls.maxCertificateChainLength=50", "-cp", tmpDir, "Curl", caPath, testCaseName + ".localhost", fmt.Sprintf("https://localhost:%d/ok", tlsPort),
		}
	})
}
